doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,321 @@
+`oneshot::Sender<Result<(), StorageError>>` that it can use to optionally notify any listeners when
+and if the updates were successfully committed.
+
+Adding the `CollectionAppender` introduces a possible correctness issue though:
+
+1. Create a `CollectionAppender` for `GlobalId(A)`.
+2. Drop source associated with `GlobalId(A)`.
+3. Try to append data.
+
+To protect against this we can change the `CollectionManager` to return an error via the
+`oneshot::Sender` if the [provided GlobalId no longer exists](https://github.com/MaterializeInc/materialize/blob/main/src/storage-client/src/controller/collection_mgmt.rs#L83).
+
+Adding this new API gets us two important properties that the existing `append` API does not have:
+
+1. The ability to send updates to the `StorageController` from a thread other than the main
+   `Coordinator` thread.
+2. Allows the `StorageController` to pick the timestamp for the update, instead of the
+   `Coordinator`, which aligns with our long term goals if having a storage owned "push source" in
+   the future.
+
+### Request Validation
+[request_validation]: #request_validation
+
+Every `WEBHOOK` source will be open to the entire internet, as such we need to include some sort of
+validation so we can verify that requests are legitimate. In general webhooks are validated by
+HMAC-ing the request body using SHA256, and validating the result with a signature provided in the
+request headers. An issue though is everyone does this just a little bit different:
+
+* Segment signature = HMAC of just the body of the request
+* Buildkite signature = HMAC of "#{timestamp}.#{body}"
+* GitHub signature = "body=" + HMAC of {body}
+
+As such we'll need to support some custom logic for validation. What we can do is support a
+`VALIDATE USING` statement that accepts only a single scalar expression. This expression will
+be provided the request `headers` as `map[text] -> text`, and the `body` as `text` _regardless of
+whether or not the source has `INCLUDE HEADERS` or what `FORMAT BODY` is specified._ For example:
+
+```
+VALIDATE USING (
+  headers['X-Signature'] = hmac('sha256', SECRET webhook_secret, body)
+)
+```
+
+Internally the provided expression will be turned into a `MirScalarExpr` which will be used to
+evaluate each request. This evaluation will happen off the main coordinator thread.
+
+### Request Limits
+[request_limits]: #request_limits
+
+To start we'll aim to support at least 100 requests per second, in other words, 10 sources could
+receive 10 requests per second. We can probably scale further, but this seems like a sensible limit
+to start at. Also we'll add an [`axum::middleware`](https://docs.rs/axum/latest/axum/middleware/index.html)
+that enforces rate limiting on our new endpoint. This should be a pretty easy way to prevent a
+misbehaving source from taking down all of `environmentd`.
+
+We'll also introduce a maximum size the body of each request can be, the default limit will be
```

Comment:
This still leaves us vulnerable to malicious requests stuffing the headers full of junk. Total request size should be capped somehow. Maybe not needed in this document or any initial work though.

Reply:
Updated to clarify the entire request size